### ISSUE TITLE
run module or package with pulumi python

### ISSUE THREAD

**Issue by Panaetius** [pulumi/pulumi#3635]
Currently, the pulumi python code uses `runpy.run_path()` for the execution, but that only works for plain script files or directories that contain a `__main__.py`.  

Additionally, if `main` in the project yaml is set to a directory, the python plugin sets the CWD to that path and then runs `.`, as opposed to running the path with the CWD being set to the directory the `Pulumi.yaml` is located in.

The docs mention that pulumi supports properly modularized python projects ( https://www.pulumi.com/docs/intro/languages/python/#getting-started ) but I don't see how that's true with the reliance on `runpy.run_path()`, as that'd require `runpy.run_module()` instead.

If  `__main__,py` is executed with CWD set to the directory it's in, then even if it resides inside a proper module, one can't use fully qualified (`module.package.file` ) or relative (`.package.file`) style imports, but instead it's only possible to import directly based on path (`package.file`) or by hacking around with pythons `sys.path`, both of which aren't great solutions. 

So from what I understand, structuring a pulumi python project as a module doesn't make sense, as it either doesn't work or it requires the module to be structured in such a way that it doesn't work as a proper module outside of usage with pulumi.

Would it be possible to extend the plugin to also support python modules as well as  python scripts? That'd be great!

And going a bit more in depth on how to structure a pulumi python project as a module in the docs would be great! Currently the docs are a bit sparse on this point.

Thanks a lot!

**Comment by bgdnlp** [comment]
#1641 might be relevant?

**Comment by Panaetius** [comment]
Yes that's almost exactly the issue I'm describing. Especially that relative imports don't work at the moment.

Another idea I had would be to support python `setuptools` entrypoints instead of depending on main (e.g. https://stackoverflow.com/a/9615473), then `Pulumi.yaml` could just point to a python package and execute the (predefined) pulumi entrypoint .

**Comment by RobbieMcKinstry** [comment]
In conjunction with https://github.com/pulumi/pulumi/issues/7360, this would be a nice QOL change for Python users.

**Comment by MprBol** [comment]
I added this piece of code in my __main__.py. It imports all .py files in the 'rules' directory. It seems to the job.

Honestly, I would have preferred to explicitly register a certain rule to the Pulumi runtime (e.g. via a decorator function or a pulumi.register() call), as opposed to whatever magic that Pulumi is using. 

```
import importlib
import pathlib
from pathlib import Path

for path in Path("rules").rglob("*.py"):
    """A bit of magic logic which automatically finds and import all defined rules in the "rules/" directory so pulumi finds them as apparently its not possible to explicitly register them to the pulumi runtime for reasons i  dont understand."""
    if "__init__.py" not in str(path) and "tests" not in str(path):
        modulename = str(path).replace("/", ".")[:-3]
        print(f"Loading rules from {modulename}")

        importlib.import_module(f".{modulename}", "myiacmodule")
```
